[lldb] handle relative working directories in swiftmodule debug info

rmaz opened the pull request:
This change makes it possible to use relative working directories. Prior
to this change the `path::append` will put a / between the `clang_argument`
and the `cur_working_dir`, which will not correctly resolve if
`cur_working_dir` is relative. This change joins the path components
separately then concatenates with the `clang_argument`.

rmaz committed:
[lldb] handle relative working directories in swiftmodule debug info

This change makes it possible to use relative working directories. Prior
to this change the `path::append` will put a / between the `clang_argument`
and the `cur_working_dir`, which will not correctly resolve if
`cur_working_dir` is relative. This change joins the path components
separately.

bulbazord commented:
@adrian-prantl @JDevlieghere @kastiglione

adrian-prantl commented:
Would you mind turning `ApplyWorkingDir` into a static function and adding a unit test to TestSwiftASTContext.cpp, so we can document the expected behavior?

kastiglione commented:
> which will not correctly resolve if `cur_working_dir` is relative

when does this happen?

rmaz commented:
> when does this happen?

At the moment it doesn't, the compiler will always prefix the current directory on a relative working directory to make it absolute. This is the first part of a series of PRs to prefix the serialized debug info in swiftmodule files. This requires treating the working directory as a relative path, as it is generally replaced with either `.` or `SRCROOT` or similar. 

I'm still verifying the upstream Swift change, but you can see most of it here: https://github.com/rmaz/swift/commit/2c5d792314d9e4a5ac2cb774cb3bfdcfceba0405

rmaz commented:
> Would you mind turning `ApplyWorkingDir` into a static function and adding a unit test to TestSwiftASTContext.cpp, so we can document the expected behavior?

Not quite sure what you mean by static in this context, but I moved the function out of the anonymous namespace and added a unit test for it. Is this what you had in mind?

kastiglione commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.h:
```diff
@@ -981,4 +981,11 @@ class SwiftASTContextForExpressions : public SwiftASTContext {
+#ifndef NDEBUG
+/// Provided only for unit tests.
+template <typename SmallString>
+void ApplyWorkingDir(SmallString &clang_argument, llvm::StringRef cur_working_dir);
```
From the looks of the function, it doesn't need to be a template, looks like `clang_argument` can be a `StringRef`.

kastiglione commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.h:
```diff
@@ -981,4 +981,11 @@ class SwiftASTContextForExpressions : public SwiftASTContext {
+#ifndef NDEBUG
+/// Provided only for unit tests.
+template <typename SmallString>
+void ApplyWorkingDir(SmallString &clang_argument, llvm::StringRef cur_working_dir);
```
c++ question: should this have an explicit instantiation? or does it not matter?

kastiglione replied:
this was discussed offline, the conclusion has been: explicit instantiation not needed

adrian-prantl replied:
This works. I originally had in mind a static member function of SwiftASTContext, but this is equivalent.

Instead of making this a template, I think we can use the base class `SmallStringImpl &` here.

adrian-prantl replied:
Actually, this doesn't work. Now release builds won't be able to build the unit test target. Can we make it a static member of Swift ASTContext instead?

rmaz replied:
This was based on the existing code further up the header:

https://github.com/apple/llvm-project/blob/swift/main/lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.h#L176-L179

Would you rather the `NDEBUG` guard is not present? I'll update the function to be a static member fn and remove the template.

adrian-prantl replied:
Yes, but the test that depends on this constructor is also guarded by a `#ifndef NDEBUG` (line 195).
If we make the function a static member of SwiftASTContext, we can leave the tests unguarded. I think that's the cleanest approach.

rmaz replied:
I don't see a `SmallStringImpl` base class anywhere though, it appears to inherit directly from `SmallVector`. Fine with keeping this templated for now?

adrian-prantl replied:
You're right, it should have been SmallVectorImpl, not SmallStringImpl. Here's an example API that works like this:
https://www.llvm.org/doxygen/namespacellvm_1_1sys_1_1fs.html#a2363fbfbd0a348f5d81fc3d3223ecae3

If using the underlying SmallVectorImpl is too cumbersome, I would just hardcode the buffer size instead of using a template. But it looks like we are only converting clang_argument to a StringRef and append characters to it.

rmaz replied:
I see, yes, that works out fine, I've updated it now.

kastiglione reviewed (commented):
Looks good to me, but let's see give @adrian-prantl a chance to look at the tests since he asked for them.

adrian-prantl commented on lldb/unittests/Symbol/TestSwiftASTContext.cpp:
```diff
@@ -220,6 +220,43 @@ TEST_F(TestSwiftASTContext, SwiftFriendlyTriple) {
+
+  single_arg_rel_path = llvm::SmallString<128>("-fmodule-map-file=module.modulemap");
+  ApplyWorkingDir(single_arg_rel_path, rel_working_dir);
+  EXPECT_EQ(single_arg_rel_path, llvm::SmallString<128>("-fmodule-map-file=rel/dir/module.modulemap"));
```
This looks great, thanks!

adrian-prantl reviewed (approved):
Thanks! Please don't forget to open a second PR that cherry-picks this patch to swift/next!

kastiglione commented:
@swift-ci test

kastiglione commented:
Thanks Richard. Could you open a matching PR against the `swift/next` branch too? thanks

rmaz commented:
> Thanks Richard. Could you open a matching PR against the `swift/next` branch too? thanks

Will do, thanks for merging.
